tmt/utils/jira.py

```diff
@@ -0,0 +1,134 @@
+import urllib.parse
+from collections.abc import Sequence
+from typing import TYPE_CHECKING, Any, Optional, Union, cast
+
+import fmf.utils
+
+import tmt.base
+import tmt.log
+import tmt.utils
+from tmt._compat.pathlib import Path
+from tmt.plugins import ModuleImporter
+
+if TYPE_CHECKING:
+    import jira
+
+
+import_jira: ModuleImporter['jira'] = ModuleImporter(  # type: ignore[valid-type]
+    'jira',
+    tmt.utils.ReportError,
+    "Install 'tmt+link-jira' to use the Jira linking.")
+
+
+def jira_link(
+        nodes: Sequence[Union['tmt.base.Test', 'tmt.base.Plan', 'tmt.base.Story']],
+        links: 'tmt.base.Links',
+        logger: tmt.log.Logger,
+        separate: bool = False) -> None:
+    """ Link the object to Jira issue and create the URL to tmt web service """
+    jira_module = import_jira(logger)
+
+    def create_url_params(tmt_object: tmt.base.Core) -> dict[str, Any]:
+        """ Prepare url parameters prefixed with tmt object type """
+        tmt_type = tmt_object.__class__.__name__.lower()
+        fmf_id = tmt_object.fmf_id
+
+        url_params: dict[str, Any] = {
+            'format': 'html',
+            f'{tmt_type}-url': fmf_id.url,
+            f'{tmt_type}-name': fmf_id.name
+            }
+
+        if fmf_id.path and isinstance(fmf_id.git_root, Path):
+            fmf_path = fmf_id.path.relative_to(fmf_id.git_root).resolve().as_posix()
+            fmf_path = fmf_path.removeprefix(fmf_id.git_root.as_posix())
+            url_params[f'{tmt_type}-path'] = fmf_path.lstrip('/')
+
+        if fmf_id.ref:
+            url_params[f'{tmt_type}-ref'] = fmf_id.ref
+
+        return url_params
+
+    def create_url(baseurl: str, url_params: dict[str, str]) -> str:
+        """ Prepare the full url from base and parameters """
+        return urllib.parse.urljoin(baseurl, '?' + urllib.parse.urlencode(url_params))
+
+    # Setup config tree, exit if config is missing
+    try:
+        config_tree = tmt.utils.Config()
+        linking_node = cast(Optional[fmf.Tree], config_tree.fmf_tree.find('/user/linking'))
+    except tmt.utils.MetadataError:
+        return
+
+    if not linking_node:
+        # Linking is not setup in config, therefore user does not want to use linking
+        return
+
+    logger.print(
+        f'Linking {fmf.utils.listed([type(node).__name__.lower() for node in nodes])}'
+        f' to Jira issue.')
+
+    issue_trackers: Any = linking_node.data.get('issue-tracker')
+
+    if not issue_trackers:
+        # Incorrectly setup config - wrong keyword
+        raise tmt.utils.GeneralError("Invalid config!")
```

Comment:
```suggestion
        raise tmt.utils.GeneralError("Invalid config: 'issue-tracker' key is missing or empty.")
```
This is a nitpick - why have a comment explaining what's wrong instead of describing it in the error?  

But here, I'm not sure about the whole "first one from the list of issue_trackers `issue_trackers)[0]`
Since the `link` feature is supposed to be exapandable and other issue-tracker types can be implemented, instead of selecting the first one from a list of issue-trackers, why not check for the type? `if tracker['type'] == 'jira' for tracker in issue_trackers` or something like that. Maybe even check if there aren't two instances.

Reply:
Yes, these `Invalid config!` lines were just quick attempt to resolve shouting `pyright` and make it happy. Since then @tkoscieln has not worked on this. This area needs polishing.

Reply:
Should be covered in 7ecd3bb7.